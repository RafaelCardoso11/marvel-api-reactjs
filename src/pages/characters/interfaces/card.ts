export interface ICard {
  id: number;
  name: string;
  thumbnail: { path: string; extension: string };
}
