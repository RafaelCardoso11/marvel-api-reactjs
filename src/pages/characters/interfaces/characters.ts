export interface ICharacters {
  id: number;
  name: string;
  thumbnail: { path: string; extension: string };
}
