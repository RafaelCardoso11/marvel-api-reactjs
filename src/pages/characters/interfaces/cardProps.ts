import { ICard } from "./card";

export interface ICardProps {
   card: ICard,
   index: number,
  }
  