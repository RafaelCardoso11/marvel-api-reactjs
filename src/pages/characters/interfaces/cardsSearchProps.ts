import { ICard } from "./card";

export interface ICardsSearchProps {
  data: {
    results: ICard[];
  };
}
