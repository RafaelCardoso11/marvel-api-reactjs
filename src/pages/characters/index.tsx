import { useDataFetcher } from "@/hooks/useDataFetcher";
import { useState } from "react";

import { ICardsSearchProps } from "./interfaces/charactersSearchProps";
import { Cards } from "./components/cards";
import { StyledContainer, StyledContainerSearch } from "./styles";

export const Characters = () => {
  const [queryParams, setQueryParams] = useState("characters");
  const [searchCharacters, setSearchCharacters] = useState<string>("");
  const { data: characters } = useDataFetcher<ICardsSearchProps>(queryParams);

  const handleSearch = async (event: React.ChangeEvent<HTMLInputElement>) => {
    const searchCharacters = event.target.value;
    setSearchCharacters(searchCharacters);
    if (searchCharacters) {
      setQueryParams("characters?nameStartsWith=" + searchCharacters);
    } else {
      setQueryParams("characters");
    }
  };

  return (
    <StyledContainer data-testid="id-container-characters">
      <StyledContainerSearch data-testid="id-container-search-characters">
        <input
          type="text"
          placeholder="Pesquisar..."
          data-testid="id-searchCharacters"
          value={searchCharacters}
          onChange={handleSearch}
        />
      </StyledContainerSearch>
      {characters && <Cards data={characters.data} data-testid="id-cards-characters"/>}
    </StyledContainer>
  );
};
