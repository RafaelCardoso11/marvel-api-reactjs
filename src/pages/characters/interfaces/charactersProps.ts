import { ICharacters } from "./characters";

export interface ICharactersProps {
   characters: ICharacters,
   index: number,
  }
  