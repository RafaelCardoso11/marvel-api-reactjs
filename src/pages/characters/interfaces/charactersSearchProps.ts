import { ICharacters } from "./characters";

export interface ICardsSearchProps {
  data: {
    results: ICharacters[];
  };
}
