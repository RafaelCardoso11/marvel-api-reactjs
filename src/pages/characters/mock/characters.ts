import { ICardsSearchProps } from "../interfaces/charactersSearchProps";

export const characters: ICardsSearchProps = {
  data: {
    results: [
      {
        id: 1011334,
        name: "3-D Man",
        thumbnail: {
          path: "http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784",
          extension: "jpg",
        },
      },
    ],
  },
};
