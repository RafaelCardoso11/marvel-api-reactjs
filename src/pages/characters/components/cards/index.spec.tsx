import { cleanup, render } from "@testing-library/react";
import { Cards } from ".";
import { charactersMock } from "@/mocks/characters";
import { ICardsSearchProps } from "../../interfaces/charactersSearchProps";
import RoutersContextTesting from "@/tests/routers";

const charactersMockEmpty: ICardsSearchProps = {
  data: {
    results: [],
  },
};

describe("<Cards/>", () => {
  afterEach(cleanup);
  afterAll(() => {
    jest.unmock("@/hooks/useDataFetcher");
  });
  beforeEach(() => {
    jest.clearAllMocks();
  });
  it("Deveria renderizar na tela um container dos Cards", () => {
    const screen = render(
      <RoutersContextTesting>
        <Cards data={charactersMockEmpty.data} />
      </RoutersContextTesting>
    );
    const cardsContainer = screen.getByTestId("id-cards-container");
    expect(cardsContainer).toBeInTheDocument();
    expect(screen.container).toMatchSnapshot();
    expect.assertions(2);
  });
  it("Deveria renderizar na tela um container com cards ao injetar os personagens", () => {
    const screen = render(
      <RoutersContextTesting>
        <Cards data={charactersMock.data} />
      </RoutersContextTesting>
    );
    const cardsContainer = screen.getByTestId("id-cards-container");

    const cards = screen.getAllByTestId("id-cards-content");

    expect(cards.length).toBe(charactersMock.data.results.length);
    expect(cardsContainer).toBeInTheDocument();
    expect(screen.container).toMatchSnapshot();
    expect.assertions(3);
  });
});
