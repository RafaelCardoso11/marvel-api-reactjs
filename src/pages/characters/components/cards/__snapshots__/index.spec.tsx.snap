// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cards/> Deveria renderizar na tela o container e os cards 1`] = `
<div>
  <section
    class="sc-gZGWnS ixooVO"
    data-testid="id-cards-container"
  >
    <div
      class="sc-kvAqsT crfTLG"
    >
      <h2
        class="sc-cdtKKh hnyVce"
      >
        3-D Man
      </h2>
      <img
        alt=""
        class="sc-grJzXs hvxsKK"
        src="http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784.jpg"
      />
    </div>
    <div
      class="sc-kvAqsT crfTLG"
    >
      <h2
        class="sc-cdtKKh hnyVce"
      >
        Hulk
      </h2>
      <img
        alt=""
        class="sc-grJzXs hvxsKK"
        src="http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784.jpg"
      />
    </div>
  </section>
</div>
`;
